src/Storages/MergeTree/MergeTreeDeduplicationLog.cpp

```diff
@@ -160,7 +160,10 @@ void MergeTreeDeduplicationLog::rotate()
     existing_logs.emplace(current_log_number, log_description);
 
     if (current_writer)
+    {
+        current_writer->finalize();
         current_writer->sync();
```

Comment:
Why are you calling `sync` after `finalize`? Why do you even need both of them? I mean mustn't `finalize()` include `sync()`?

Reply:
> Why are you calling `sync` after `finalize`? Why do you even need both of them? I mean mustn't `finalize()` include `sync()`?

Well. This is tricky.  sync is meaningless for S3. finalize is crucial  for S3.
That interface is written and broadly used with files. For files there are different priorities. With files exactly such approach is used. I was trying to make it consistent with other use cases. For files finalize do final operation in buffer to write the file, sync calls syscall for flush the filesystem buffers. And sync is called exactly after finish.

Reply:
Maybe we can make `finalize()` call `sync()` for files?

Reply:
I'm not sure that it is good idea. For files sync is omitted usually. Sync is called only if it is strictly necessary or under a settings/flag. This is done due to performance.

Reply:
`finalize()` is usually the last thing we do with any write buffer. Could we call `finalize()` after `sync()`?